### PR TITLE
[Unification] Unified text embeddings

### PR DESCRIPTION
**Pull request by RdoubleA**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #253
* #249
* __->__ #248
* #244
* #243
* #240
* #239

## Summary
Create a unified `TextEmbeddings` class with word, position, token type embeddings + layernorm + dropout. 
- Add flag to offset position ids by pad token id like in RoBERTa for the `MDETRTextEmbeddings`. 
- Refactor ALBEF and MDETR with unified class
- Refactoring FLAVA will come in a later PR since the checkpoint must be updated.

## Test plan
`pytest test/modules/layers/test_text_embedding.py -vv`
```
===================================================================================== test session starts ======================================================================================
platform linux -- Python 3.9.12, pytest-7.1.1, pluggy-1.0.0 -- /fsx/users/rafiayub/conda/envs/torchmm/bin/python
cachedir: .pytest_cache
rootdir: /data/home/rafiayub/torchmultimodal, configfile: pyproject.toml
plugins: hydra-core-1.1.2, cov-3.0.0, mock-3.8.2
collected 3 items                                                                                                                                                                              

test/modules/layers/test_text_embedding.py::TestTextEmbeddings::test_forward PASSED                                                                                                      [ 33%]
test/modules/layers/test_text_embedding.py::TestTextEmbeddings::test_invalid_input PASSED                                                                                                [ 66%]
test/modules/layers/test_text_embedding.py::TestTextEmbeddings::test_create_position_ids_from_input_ids PASSED                                                                           [100%]

====================================================================================== 3 passed in 1.31s =======================================================================================
```
`pytest test/models/albef -vv`
```
===================================================================================== test session starts ======================================================================================
platform linux -- Python 3.9.12, pytest-7.1.1, pluggy-1.0.0 -- /fsx/users/rafiayub/conda/envs/torchmm/bin/python
cachedir: .pytest_cache
rootdir: /data/home/rafiayub/torchmultimodal, configfile: pyproject.toml
plugins: hydra-core-1.1.2, cov-3.0.0, mock-3.8.2
collected 23 items                                                                                                                                                                             

test/models/albef/test_albef.py::test_albef_image_embeddings PASSED                                                                                                                      [  4%]
test/models/albef/test_albef.py::test_albef_image_embeddings_momentum PASSED                                                                                                             [  8%]
test/models/albef/test_albef.py::test_albef_text_embeddings PASSED                                                                                                                       [ 13%]
test/models/albef/test_albef.py::test_albef_text_embeddings_momentum PASSED                                                                                                              [ 17%]
test/models/albef/test_albef.py::test_albef_multimodal_embeddings PASSED                                                                                                                 [ 21%]
test/models/albef/test_albef.py::test_albef_multimodal_embeddings_momentum PASSED                                                                                                        [ 26%]
test/models/albef/test_albef.py::test_copy_params_momentum_models PASSED                                                                                                                 [ 30%]
test/models/albef/test_albef.py::test_dequeue_and_enqueue PASSED                                                                                                                         [ 34%]
test/models/albef/test_albef.py::test_momentum_update PASSED                                                                                                                             [ 39%]
test/models/albef/test_albef.py::test_similarity PASSED                                                                                                                                  [ 43%]
test/models/albef/test_albef.py::test_neg_embeddings PASSED                                                                                                                              [ 47%]
test/models/albef/test_image_encoder.py::TestALBEFVisionEncoder::test_vision_transformer PASSED                                                                                          [ 52%]
test/models/albef/test_image_encoder.py::TestALBEFVisionEncoder::test_invalid_input_length PASSED                                                                                        [ 56%]
test/models/albef/test_image_encoder.py::TestALBEFVisionEncoder::test_invalid_image_channel_dim PASSED                                                                                   [ 60%]
test/models/albef/test_image_encoder.py::TestALBEFVisionEncoder::test_invalid_image_height PASSED                                                                                        [ 65%]
test/models/albef/test_image_encoder.py::TestALBEFVisionEncoder::test_invalid_image_width PASSED                                                                                         [ 69%]
test/models/albef/test_multimodal_encoder.py::test_multimodal_encoder PASSED                                                                                                             [ 73%]
test/models/albef/test_multimodal_encoder.py::test_invalid_image_hidden_size PASSED                                                                                                      [ 78%]
test/models/albef/test_multimodal_encoder.py::test_invalid_text_hidden_size PASSED                                                                                                       [ 82%]
test/models/albef/test_multimodal_encoder.py::test_not_matching_input_batch_size PASSED                                                                                                  [ 86%]
test/models/albef/test_text_encoder.py::test_text_encoder PASSED                                                                                                                         [ 91%]
test/models/albef/test_text_encoder.py::test_invalid_input_length PASSED                                                                                                                 [ 95%]
test/models/albef/test_text_encoder.py::test_not_matching_attention_mask_shape PASSED                                                                                                    [100%]

====================================================================================== 23 passed in 3.18s ======================================================================================
```
`pytest test/models/mdetr -vv`
```
===================================================================================== test session starts ======================================================================================
platform linux -- Python 3.9.12, pytest-7.1.1, pluggy-1.0.0 -- /fsx/users/rafiayub/conda/envs/torchmm/bin/python
cachedir: .pytest_cache
rootdir: /data/home/rafiayub/torchmultimodal, configfile: pyproject.toml
plugins: hydra-core-1.1.2, cov-3.0.0, mock-3.8.2
collected 7 items                                                                                                                                                                              

test/models/mdetr/test_image_encoder.py::TestMDETRImageEncoder::test_resnet_101_forward PASSED                                                                                           [ 14%]
test/models/mdetr/test_mdetr.py::TestMDETR::test_mdetr_model PASSED                                                                                                                      [ 28%]
test/models/mdetr/test_text_encoder.py::TestMDETRTextEncoder::test_mdetr_modified_transformer PASSED                                                                                     [ 42%]
test/models/mdetr/test_text_encoder.py::TestMDETRTextEncoder::test_mdetr_text_embeddings PASSED                                                                                          [ 57%]
test/models/mdetr/test_text_encoder.py::TestMDETRTextEncoder::test_mdetr_text_encoder PASSED                                                                                             [ 71%]
test/models/mdetr/test_transformer.py::TestMDETRTransformer::test_transformer_encoder PASSED                                                                                             [ 85%]
test/models/mdetr/test_transformer.py::TestMDETRTransformer::test_transformer_decoder PASSED                                                                                             [100%]

====================================================================================== 7 passed in 5.59s =======================================================================================
```

Differential Revision: [D38551546](https://our.internmc.facebook.com/intern/diff/D38551546)

### CODE REVIEW

File: torchmultimodal/modules/layers/text_embedding.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import torch
+from torch import nn, Tensor
+
+
+class TextEmbeddings(nn.Module):
+    """Construct word, position, and token type embeddings following BERT.
+
+    Attributes:
+        hidden_size (int): size of embedding space. Default is 768.
+        vocab_size (int): size of vocabulary. Default is 30522.
```
Comment (langong347) [inline torchmultimodal/modules/layers/text_embedding.py:18]: If this is supposed to be a horizontal components, do we still need to set default values (I assume they all come from a specific model right?) So to be clear to the users, we could either:

- Defer those default values to the model that uses this component and keeping it non-opinionated here.
- Rename this as `BERTTextEmbeddings` to avoid over generalization

**Review by langong347 — APPROVED** [review]

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38551546).

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38551546).

File: test/models/mdetr/test_text_encoder.py [hunk 2]
```diff
@@ -10,25 +10,26 @@
...
-        self.embeddings = MDETRTextEmbeddings(
+        self.embeddings = BERTTextEmbeddings(
             hidden_size=self.hidden_size,
             vocab_size=50265,
             pad_token_id=1,
             type_vocab_size=1,
             max_position_embeddings=self.max_position_embeddings,
             layer_norm_eps=1e-05,
-            hidden_dropout_prob=0.1,
+            dropout=0.1,
```
Comment (ankitade) [inline test/models/mdetr/test_text_encoder.py:31]: why setting dropout in test? remove

File: test/modules/layers/test_text_embedding.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+@pytest.fixture(autouse=True)
+def random():
+    set_rng_seed(4)
+
+
+class TestTextEmbeddings:
+    @pytest.fixture
+    def text_embedding(self):
+        return BERTTextEmbeddings(hidden_size=3, vocab_size=3)
```
Comment (ankitade) [inline test/modules/layers/test_text_embedding.py:21]: add .eval?

File: torchmultimodal/modules/layers/text_embedding.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        max_position_embeddings: int = 512,
+        type_vocab_size: int = 2,
+        layer_norm_eps: float = 1e-12,
+        dropout: float = 0.0,
+        offset_pos_ids: bool = False,
+    ) -> None:
+        super().__init__()
+        self.word_embeddings = nn.Embedding(vocab_size, hidden_size, pad_token_id)
+        self.position_embeddings = nn.Embedding(max_position_embeddings, hidden_size)
+        self.token_type_embeddings = nn.Embedding(type_vocab_size, hidden_size)
```
Comment (ankitade) [inline torchmultimodal/modules/layers/text_embedding.py:50]: can u add a comment that this is not used in original roberta but keeping here for parity with HF checkpoints

**Review by ankitade — APPROVED** [review]

File: test/models/mdetr/test_text_encoder.py (hunk 2 above)
Comment (RdoubleA) [reply test/models/mdetr/test_text_encoder.py:31]: this was originally there, I'm inclined to leave as is so I don't have to update the test values

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38551546).
